debian/unclutter/usr/share/man/man1/unclutter.1
.\"unclutter man
.TH UNCLUTTER 1
.SH NAME
unclutter \- remove idle cursor image from screen
.SH SYNOPSIS
.B
unclutter
.RB [ -display | -d
.IR display ]
.RB [ -idle
.IR seconds ]
.RB [ -keystroke ]
.RB [ -jitter
.IR pixels ]
.RB [ -grab ]
.RB [ -noevents ]
.RB [ -reset ]
.RB [ -root ]
.RB [ -onescreen ]
.RB [ -visible ]
.RB [ -regex ]
.RB [ -not | -notname
.IR name " " ... ]
.RB [ -notclass
.IR class " " ... ]
.SH DESCRIPTION
.B unclutter
removes the cursor image from the screen so that it does not
obstruct the area you are looking at after it has not moved for
a given time.
It does not do this if the cursor is in the root window or a button is down.
It tries to ignore jitter (small movements due to noise)
if you have a mouse that twitches.
.SH OPTIONS
.TP
-display
is followed by the display to open.
.TP
-idle
is followed by the number of seconds between polls for idleness.
The default is 5. Supports subsecond idle times.
.TP
-keystroke
tells
.B unclutter
not to use a timeout to determine when to remove the cursor, but to
instead wait until a key has been pressed (released, really).
.TP
-jitter
is followed by the amount of movement of the pointer that is to be ignored
and considered as random noise.
The default is 0.
.TP
-grab
means use the original method of grabbing the pointer in order to remove the
cursor. 
This often doesn't interoperate too well with some window managers.
.TP
-noevents
stops unclutter sending a pseudo EnterNotify event to the X client whose
cursor has been stolen.
Sending the event helps programs like emacs think that they have not
lost the pointer focus.
This option is provided for backwards compatibility in case some
clients get upset.
.TP
-reset
resets the timeout for idleness after the cursor is restored for some reason
(such as a window being pushed or popped) even though the x y coordinates
of the cursor have not changed.
Normally, the cursor would immediately be removed again.
.TP
-root
means remove the cursor even if it is on the root background, where in
principle it should not be obscuring anything useful.
.TP
-onescreen
restricts unclutter to the single screen specified as display,
or the default screen for the display.
Normally, unclutter will unclutter all the screens on a display.
.TP
-visible
ignore visibility events (does not apply to -grab).
If the cursor never gets hidden, despite a generous -jitter value,
try this option
.TP
-not
is followed by a list of window names where the cursor should not be
removed.
The first few characters of the WM_NAME property on the window need
to match one the listed names.
This argument must be the last on the command line.
.TP
-notname
is exactly the same as -not
.TP
-notclass
is similar to -notname, except that the WM_CLASS property of the window is used.
This argument must be the last on the command line, and so cannot be used
with -not or -notname.
.TP
-regex
treats the first name or class (see above) as a regular expression.
This means that `` -regex -not foo bar '' will not work as expected; instead
use `` -regex -not 'foo|bar' ''.
.SH LIMITATIONS
The -keystroke option may not work (that is, the cursor will not
disappear) with clients that request KeyRelease events.
Games and Xt applications using KeyUp in their translation tables are 
most likely to suffer from this problem.
The most feasible solution is to extend unclutter to use the XTest
extension to get all keyboard and mouse events, though this of course
requires XTest to be in the server too.
.PP
The -keystroke option does not distinguish modifier keys from
keys which actually generate characters.
If desired this could be implemented in a simple way by using XLookupString
to see if any characters are returned.
.SH DIAGNOSTICS
The message
.sp
 someone created a sub-window to my sub-window!
.sp 
means that unclutter thinks a second unclutter is running, and tried
to steal the cursor by creating a sub-window to the sub-window already
used to steal the cursor.
This situation quickly deteriorates into a fight no one can win, so
it is detected when possible and the program gives up.
.SH AUTHOR
Mark M Martin. cetia 7feb1994. mmm@cetia.fr


debian/changelog
unclutter (8-12) unstable; urgency=low

  * Updating maintainer field.
  * Updating vcs fields.
  * Not calling debconf-updatepo in clean target of rules anymore.
  * Adding maintainer homepage in copyright.
  * Using quilt to manage upstream changes.
  * Correcting wrong vcs-browser field.
  * Adding explicit debian source version 1.0 until switch to 3.0.

 -- Daniel Baumann <daniel@debian.org>  Wed, 16 Dec 2009 22:23:23 +0100

unclutter (8-11) unstable; urgency=low

  * Using correct rfc-2822 date formats in changelog.
  * Updating year in copyright file.
  * Removing manual addition of debconf in depends.
  * Using more common directory name for storing local debian addisions.
  * Adding Russian debconf translations from Yuri Kozlov
    <yuray@komyakino.ru> (Closes: #541750).
  * Updating package to standards version 3.8.3.
  * Sorting depends.
  * Minimizing rules file.

 -- Daniel Baumann <daniel@debian.org>  Sat, 22 Aug 2009 14:12:22 +0200

unclutter (8-10) unstable; urgency=low

  * Prefixing debhelper files with package name.
  * Updating package to standards 3.8.1.
  * Updating copyright for debian files.

 -- Daniel Baumann <daniel@debian.org>  Sun, 12 Apr 2009 10:13:00 +0200

unclutter (8-9) unstable; urgency=medium

  * Updating vcs fields in control file.
  * Adding Spanish debconf translations from Francisco Javier Cuadrado
    <fcocuadrado@gmail.com> (Closes: #499822).

 -- Daniel Baumann <daniel@debian.org>  Tue, 23 Sep 2008 07:25:00 +0200

unclutter (8-8) unstable; urgency=low

  * Correcting typo in header of ja.po, thanks to Christian Perrier
    <bubulle@debian.org> (Closes: #490928).

 -- Daniel Baumann <daniel@debian.org>  Tue, 15 Jul 2008 14:12:00 +0200

unclutter (8-7) unstable; urgency=low

  * New maintainer (Closes: #489451).
  * Upgrading package to standards 3.8.0.
  * Updating vcs fields in control file.
  * Rewriting copyright file in machine-interpretable format.
  * Reverting rules to long format since I prefere this.
  * Using dedicated debhelper files rather than arguments to dh_* calls,
    I do prefere this as well.
  * Adding default and Xsession files, using debconf to populate it
    (Closes: #483055).

 -- Daniel Baumann <daniel@debian.org>  Mon, 14 Jul 2008 07:57:00 +0200

unclutter (8-6) unstable; urgency=low

  * debhelper v7, rules file minimisation
  * Orphaned the package (again).

 -- Joey Hess <joeyh@debian.org>  Sat, 05 Jul 2008 17:33:13 -0400

unclutter (8-5) unstable; urgency=low

  * Document -visible in man page. Thanks, Marc-Jano Knopp. Closes: #384365
  * Fixed a couple typos in help text.

 -- Joey Hess <joeyh@debian.org>  Thu, 24 Aug 2006 16:41:35 -0400

unclutter (8-4) unstable; urgency=low

  * Patch from Jonas Koelker to support subsecond idle times. Closes: #369891
  * debhelper v5
  * Current standards version.

 -- Joey Hess <joeyh@debian.org>  Sun, 11 Jun 2006 16:51:08 -0400

unclutter (8-3) unstable; urgency=low

  * Remove the menu file, I think the changes that anyone will want to
    start/stop unclutter from the menu is slim and its use of killall was
    evil.
  * Current standards version.

 -- Joey Hess <joeyh@debian.org>  Sun, 18 Dec 2005 17:24:23 -0500

unclutter (8-2) unstable; urgency=low

  * Rebuild to use new split X library packages, update build depends.
    Closes: #287675
  * Menu file updates (qouting).

 -- Joey Hess <joeyh@debian.org>  Wed, 29 Dec 2004 15:51:24 -0500

unclutter (8-1) unstable; urgency=low

  * Taking this package over again.
  * Update to current policy. Closes: #197714
  * Changed version number to 8, since the original tarball is named
    unclutter-8.tar.Z.
  * Downloaded original tarball, and included it as pristine upstream source,
    except I had to uncompress it (.Z) and gzip it back up. Still, more
    pristine than before.
  * Include patchlevel.h as the upstream changelog, such as it is.
  * Expanded copyright file with download location and actual non-copyright
    statement.
  * Rewrote and simplified the package description.
  * Update to debhelper v4.
  * Moved out of the X directories since the Imakefile is not actually used.
  * Patched it to not spin when a full-screen program (such as a SDL game
    or xlock) is running and the XGrabPointer fails. Closes: #197684
  * Applied patch from Stelios Bounanos that adds a -notclass switch, useful
    for apps like galeon that frequently change the window title.
    Closes: #149720
  * Also a patch to add regex support in the -not matching.

 -- Joey Hess <joeyh@debian.org>  Tue, 17 Jun 2003 10:54:44 -0400

unclutter (0.8-22) unstable; urgency=low

  * Updated build-deps to use new xlibs-dev (Closes: #99691)
  * Removed dh_suidregister call from debian/rules

 -- Bradley Marshall <bmarshal@pisoftware.com>  Sun, 03 Jun 2001 21:19:58 +1000

unclutter (0.8-21) unstable; urgency=low

  * Change of maintainer

 -- Bradley Marshall <bmarshal@pisoftware.com>  Tue, 22 Aug 2000 19:37:33 +1000

unclutter (0.8-20) unstable; urgency=low

  * Build deps.

 -- Joey Hess <joeyh@debian.org>  Sat, 04 Dec 1999 16:42:27 -0800

unclutter (0.8-19) unstable; urgency=low

  * FHS

 -- Joey Hess <joeyh@debian.org>  Mon, 06 Sep 1999 16:56:50 -0700

unclutter (0.8-18) unstable; urgency=low

  * Fix minior lintian errors.

 -- Joey Hess <joeyh@debian.org>  Sat, 14 Mar 1998 20:05:24 -0800

unclutter (0.8-17) unstable; urgency=low

  * Updated standards-version.

 -- Joey Hess <joeyh@debian.org>  Mon, 09 Feb 1998 15:09:57 -0800

unclutter (0.8-16) unstable; urgency=low

  * Use debhelper.

 -- Joey Hess <joeyh@debian.org>  Sat, 07 Feb 1998 20:03:20 -0800

unclutter (0.8-15) unstable; urgency=low

  * libc6.
  * Routine update of debian/rules:
    Fixed binary-indep target.

 -- Joey Hess <joeyh@debian.org>  Mon, 08 Sep 1997 19:31:27 -0400

unclutter (0.8-14) unstable; urgency=low

  * Use debstd.

 -- Joey Hess <joeyh@debian.org>  Mon, 14 Apr 1997 13:46:51 -0400

unclutter (0.8-13) frozen unstable; urgency=low

  * New maintainer.

unclutter.c
/*
 * mousegrab: lock mouse to initial window and hide it.
 * right mouse button to get it back.
 *
 * Mouse will be hidden but buttons and mouse wheel still work,
 * and app can receive keystrokes.
 *
 * useful if you have an optical mouse that jitters when off the floor
 * and you'd rather not have it wander all over the place, or you need
 * to click on something without worrying about mouse being moved ...
 * Especially nice to use with easystroke -> start with mouse gesture !
 * 
 * implementation: uses XGrabPointer to grab the mouse until released,
 * so focus etc should pretty much stay where they were when it started.
 * button events are forwarded to initial window.
 *
 * based on unclutter code, by Mark M Martin.
 *
 * lemonsqueeze <lemonsqueeze@gmx.com>
 */

/*
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

#include <X11/Xos.h>
#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <X11/Xproto.h>
#include <stdio.h>
#include "vroot.h"
#include <string.h>
#include <assert.h>
#include <stdlib.h>
#include <sys/types.h>
#include <regex.h>

char *progname;
pexit(str)char *str;{
    fprintf(stderr,"%s: %s\n",progname,str);
    exit(1);
}
usage(){
    pexit("usage:  moustrap\n\n\
  lock mouse to initial window and hide it.\n\
  right mouse button to get it back.\n\
  \n\
  Mouse will be hidden but buttons and mouse wheel still work,\n\
  and app can receive keystrokes.\n\
  \n\
  useful if you have an optical mouse that jitters when off the floor\n\
  and you'd rather not have it wander all over the place, or you need\n\
  to click on something without worrying about mouse being moved ...\n\
  Especially nice to use with easystroke -> start with mouse gesture !\n\
  ");
#if 0
    pexit("usage:\n\
	-display <display>\n\
	-idle <seconds>		time between polls to detect idleness.\n\
	-keystroke		wait for keystroke before idling.\n\
	-jitter <pixels>	pixels mouse can twitch without moving\n\
	-grab			use grabpointer method not createwindow\n\
	-reset			reset the timer whenever cursor becomes\n\
					visible even if it hasn't moved\n\
 	-root	       		apply to cursor on root window too\n\
	-onescreen		apply only to given screen of display\n\
 	-visible       		ignore visibility events\n\
 	-noevents      		don't send pseudo events\n\
	-regex			name or class below is a regular expression\n\
	-not names...		don't apply to windows whose wm-name begins.\n\
				(must be last argument)\n\
	-notname names...	same as -not names...\n\
	-notclass classes...    don't apply to windows whose wm-class begins.\n\
				(must be last argument, cannot be used with\n\
				-not or -notname)");
#endif
}

static void dsleep(float t)
{
    struct timeval tv;
    assert(t >= 0);
    tv.tv_sec = (int) t;
    tv.tv_usec = (t - tv.tv_sec) * 1000000;
    select(0, NULL, NULL, NULL, &tv);
}

#define ALMOSTEQUAL(a,b) (abs(a-b)<=jitter)
#define ANYBUTTON (Button1Mask|Button2Mask|Button3Mask|Button4Mask|Button5Mask)

/* Since the small window we create is a child of the window the pointer is
 * in, it can be destroyed by its adoptive parent.  Hence our destroywindow()
 * can return an error, saying it no longer exists.  Similarly, the parent
 * window can disappear while we are trying to create the child. Trap and
 * ignore these errors.
 */
int (*defaulthandler)();
int errorhandler(display,error)
Display *display;
XErrorEvent *error;
{
    if(error->error_code!=BadWindow)
	(*defaulthandler)(display,error);
}

char **names = 0;	/* -> argv list of names to avoid */
char **classes = 0;     /* -> argv list of classes to avoid */
regex_t *nc_re = 0;     /* regex for list of classes/names to avoid */

/*
 * return true if window has a wm_name (class) and the start of it matches
 * one of the given names (classes) to avoid
 */
nameinlist(display,window)
Display *display;
Window window;
{
    char **cpp;
    char *name = 0;

    if(names)
	XFetchName (display, window, &name);
    else if(classes){
	XClassHint *xch = XAllocClassHint();
	if(XGetClassHint (display, window, xch))
	    name = strdup(xch->res_class);
	if(xch)
	    XFree(xch);
    }else
	return 0;

    if(name){
	if(nc_re){
	    if(!regexec(nc_re, name, 0, 0, 0)) {
		XFree(name);
		return 1;
	    }
	}else{
	    for(cpp = names!=0 ? names : classes;*cpp!=0;cpp++){
		if(strncmp(*cpp,name,strlen(*cpp))==0)
		    break;
	    }
	    XFree(name);
	    return(*cpp!=0);
	}
    }
    return 0;
}	
/*
 * create a small 1x1 curssor with all pixels masked out on the given screen.
 */
createnullcursor(display,root)
Display *display;
Window root;
{
    Pixmap cursormask;
    XGCValues xgc;
    GC gc;
    XColor dummycolour;
    Cursor cursor;

    cursormask = XCreatePixmap(display, root, 1, 1, 1/*depth*/);
    xgc.function = GXclear;
    gc =  XCreateGC(display, cursormask, GCFunction, &xgc);
    XFillRectangle(display, cursormask, gc, 0, 0, 1, 1);
    dummycolour.pixel = 0;
    dummycolour.red = 0;
    dummycolour.flags = 04;
    cursor = XCreatePixmapCursor(display, cursormask, cursormask,
	      &dummycolour,&dummycolour, 0,0);
    XFreePixmap(display,cursormask);
    XFreeGC(display,gc);
    return cursor;
}

main(argc,argv)char **argv;{
    Display *display;
    int screen,oldx = -99,oldy = -99,numscreens;
    int doroot = 0, jitter = 0, usegrabmethod = 1, waitagain = 0,
        dovisible = 1, doevents = 1, onescreen = 0;
    float idletime = 0.0;
    Cursor *cursor;
    Window *realroot;
    Window root;
    char *displayname = 0;
    
    progname = *argv;
    argc--;
    while(argv++,argc-->0){
#if 0    	
	if(strcmp(*argv,"-idle")==0){
	    argc--,argv++;
	    if(argc<0)usage();
	    idletime = atof(*argv);
	}else if(strcmp(*argv,"-keystroke")==0){
	    idletime = -1;
	}else if(strcmp(*argv,"-jitter")==0){
	    argc--,argv++;
	    if(argc<0)usage();
	    jitter = atoi(*argv);
	}else if(strcmp(*argv,"-noevents")==0){
	    doevents = 0;
	}else if(strcmp(*argv,"-root")==0){
	    doroot = 1;
	}else if(strcmp(*argv,"-grab")==0){
	    usegrabmethod = 1;
	}else if(strcmp(*argv,"-reset")==0){
	    waitagain = 1;
	}else if(strcmp(*argv,"-onescreen")==0){
	    onescreen = 1;
	}else if(strcmp(*argv,"-visible")==0){
	    dovisible = 0;
	}else if(strcmp(*argv,"-regex")==0){
	    nc_re = (regex_t *)malloc(sizeof(regex_t));
	}else if(strcmp(*argv,"-not")==0 || strcmp(*argv,"-notname")==0){
	    /* take rest of srg list */
	    names = ++argv;
	    if(*names==0)names = 0;	/* no args follow */
	    argc = 0;
	}else if(strcmp(*argv,"-notclass")==0){
	    /* take rest of arg list */
	    classes = ++argv;
	    if(*classes==0)classes = 0;	/* no args follow */
	    argc = 0;
	}else if(strcmp(*argv,"-display")==0 || strcmp(*argv,"-d")==0){
	    argc--,argv++;
	    if(argc<0)usage();
	    displayname = *argv;
	}else
#endif
	    usage();
    }
    /* compile a regex from the first name or class */
    if(nc_re){
	if(names || classes){
	    if (regcomp(nc_re, (names != 0 ? *names : *classes),
			REG_EXTENDED | REG_NOSUB)) { /* error */
		free(nc_re);
		names = classes = 0;
		nc_re = 0;
	    }
	}else{ /* -regex without -not... ... */
	    free(nc_re);
	    nc_re = 0;
	}
    }

    display = XOpenDisplay(displayname);
    if(display==0)pexit("could not open display");
    numscreens = ScreenCount(display);
    cursor = (Cursor*) malloc(numscreens*sizeof(Cursor));
    realroot = (Window*) malloc(numscreens*sizeof(Window));

    /* each screen needs its own empty cursor.
     * note each real root id so can find which screen we are on
     */
    for(screen = 0;screen<numscreens;screen++)
	if(onescreen && screen!=DefaultScreen(display)){
	    realroot[screen] = -1;
	    cursor[screen] = -1;
	}else{
	    realroot[screen] = XRootWindow(display,screen);
	    cursor[screen] = createnullcursor(display,realroot[screen]);
	    if(idletime<0)
		XSelectInput(display,realroot[screen],KeyReleaseMask);
	}
    screen = DefaultScreen(display);
    root = VirtualRootWindow(display,screen);

    if(!usegrabmethod)
	defaulthandler = XSetErrorHandler(errorhandler);
    /*
     * create a small unmapped window on a screen just so xdm can use
     * it as a handle on which to killclient() us.
     */
    XCreateWindow(display, realroot[screen], 0,0,1,1, 0, CopyFromParent,
		 InputOutput, CopyFromParent, 0, (XSetWindowAttributes*)0);

    while(1){
	Window dummywin,windowin,topwin,newroot;
	int rootx,rooty,winx,winy, res;
	unsigned int modifs;
	Window lastwindowavoided = None;
	
	/*
	 * wait for pointer to not move and no buttons down
	 * or if triggered by keystroke check no buttons down
	 */
	while(1){
	    if(idletime<0){		/* wait for keystroke trigger */
		XEvent event;
		do{
		    XNextEvent(display,&event);
		}while(event.type != KeyRelease ||
		       (event.xkey.state & ANYBUTTON));
		oldx = event.xkey.x_root;
		oldy = event.xkey.y_root;
	    }
	    res = XQueryPointer(display, root, &newroot, &windowin,
				&rootx, &rooty, &winx, &winy, &modifs);
	    topwin = windowin;
	    if (!res)
	    {
		/* window manager with virtual root may have restarted
		 * or we have changed screens */
		if(!onescreen){
		    for(screen = 0;screen<numscreens;screen++)
			if(newroot==realroot[screen])break;
		    if(screen>=numscreens)
			pexit("not on a known screen");
		}
		root = VirtualRootWindow(display,screen);
	    }else if((!doroot && windowin==None) || (modifs & ANYBUTTON) ||
		     !(ALMOSTEQUAL(rootx,oldx) && ALMOSTEQUAL(rooty,oldy))){
		oldx = rootx, oldy = rooty;
	    }else if(windowin==None){
		windowin = root;
		break;
	    }else if(windowin!=lastwindowavoided){
		/* descend tree of windows under cursor to bottommost */
		Window childin;
		int toavoid = xFalse;
		lastwindowavoided = childin = windowin;
		do{
		    windowin = childin;
		    if(nameinlist (display, windowin)){
			toavoid = xTrue;
			break;
		    }
		}while(XQueryPointer(display, windowin, &dummywin,
		     &childin, &rootx, &rooty, &winx, &winy, &modifs)
		       && childin!=None);
		if(!toavoid){
		    lastwindowavoided = None;
		    break;
		}
	    }
	    if(idletime>=0)
		dsleep(idletime);
	}
	// printf("topwin: %x windowin: %x\n", (int)topwin, (int) windowin);
	
	/* wait again next time */
	if(waitagain)
	    oldx = -1-jitter;
	if(usegrabmethod){
	    if(XGrabPointer(display, root, 0,
		    PointerMotionMask|ButtonPressMask|ButtonReleaseMask,
		    GrabModeAsync, GrabModeAsync, None, cursor[screen],
		    CurrentTime)==GrabSuccess){
		XEvent event;
		do{
		    //Window target = topwin;    // lock on first topwin
		    Window target = windowin;
		    //Window target = event.xbutton.subwindow;

		    // TODO: exit if target window gets destroyed...
		    
		    XNextEvent(display,&event);

		    // Forward button clicks to initial window
		    if ((event.type == ButtonPress ||
			 event.type == ButtonRelease) &&
			event.xbutton.button != 3)
		    {
			event.xbutton.window = target;
			event.xbutton.subwindow = target;
			event.xbutton.x = winx;
			event.xbutton.y = winy;
			event.xbutton.x_root = rootx;
			event.xbutton.y_root = rooty;
			(void)XSendEvent(display, target,
					 1,
					 (event.type == ButtonPress ? ButtonPressMask : ButtonReleaseMask),
					 &event);
		    }

		}while(!(event.type == ButtonRelease &&
			     event.xbutton.button == 3));
		/* keep grabbing the mouse until right click */

		XUngrabPointer(display, CurrentTime);
		exit(0);

	    }else{
		/* go to sleep to prevent tight loops */
		if(idletime>=0)
			dsleep(idletime);
	    }
	}else{
	    XSetWindowAttributes attributes;
	    XEvent event;
	    Window cursorwindow;
	    
	    /* create small input-only window under cursor
	     * as a sub window of the window currently under the cursor
	     */
	    attributes.event_mask = LeaveWindowMask |
			EnterWindowMask |
			StructureNotifyMask |
			FocusChangeMask;
	    if(dovisible)
		attributes.event_mask |= VisibilityChangeMask;
	    attributes.override_redirect = True;
	    attributes.cursor = cursor[screen];
	    cursorwindow = XCreateWindow
		(display, windowin,
		 winx-jitter, winy-jitter,
		 jitter*2+1, jitter*2+1, 0, CopyFromParent,
		 InputOnly, CopyFromParent, 
		 CWOverrideRedirect | CWEventMask | CWCursor,
		 &attributes);
	    /* discard old events for previously created windows */
	    XSync(display,True);
	    XMapWindow(display,cursorwindow);
	    /*
	     * Dont wait for expose/map cos override and inputonly(?).
	     * Check that created window captured the pointer by looking
	     * for inevitable EnterNotify event that must follow MapNotify.
	     * [Bug fix thanks to Charles Hannum <mycroft@ai.mit.edu>]
	     */
	    XSync(display,False);
	    if(!XCheckTypedWindowEvent(display, cursorwindow, EnterNotify,
				      &event))
		oldx = -1-jitter;	/* slow down retry */
	    else{
		if(doevents){
		    /*
		     * send a pseudo EnterNotify event to the parent window
		     * to try to convince application that we didnt really leave it
		     */
		    event.xcrossing.type = EnterNotify;
		    event.xcrossing.display = display;
		    event.xcrossing.window = windowin;
		    event.xcrossing.root = root;
		    event.xcrossing.subwindow = None;
		    event.xcrossing.time = CurrentTime;
		    event.xcrossing.x = winx;
		    event.xcrossing.y = winy;
		    event.xcrossing.x_root = rootx;
		    event.xcrossing.y_root = rooty;
		    event.xcrossing.mode = NotifyNormal;
		    event.xcrossing.same_screen = True;
		    event.xcrossing.focus = True;
		    event.xcrossing.state = modifs;
		    (void)XSendEvent(display,windowin,
				     True/*propagate*/,EnterWindowMask,&event);
		}
		/* wait till pointer leaves window */
		do{
		    XNextEvent(display,&event);
		}while(event.type!=LeaveNotify &&
		       event.type!=FocusOut &&
		       event.type!=UnmapNotify &&
		       event.type!=ConfigureNotify &&
		       event.type!=CirculateNotify &&
		       event.type!=ReparentNotify &&
		       event.type!=DestroyNotify &&
		       (event.type!=VisibilityNotify ||
			event.xvisibility.state==VisibilityUnobscured)
		       );
		/* check if a second unclutter is running cos they thrash */
		if(event.type==LeaveNotify &&
		   event.xcrossing.window==cursorwindow &&
		   event.xcrossing.detail==NotifyInferior)
		    pexit("someone created a sub-window to my sub-window! giving up");
	    }
	    XDestroyWindow(display, cursorwindow);
	}
    }
}
